drivers/soundwire/intel_shim2.c

```diff
@@ -155,8 +156,52 @@ static void intel_debugfs_init(struct sdw_intel *sdw) {}
 static void intel_debugfs_exit(struct sdw_intel *sdw) {}
 #endif /* CONFIG_DEBUG_FS */
 
+static int intel_link_power_up(struct sdw_intel *sdw)
+{
+	int ret;
+
+	mutex_lock(sdw->link_res->shim_lock);
+
+	ret = hdac_bus_eml_sdw_power_up_unlocked(sdw->link_res->hbus, sdw->instance);
+	if (ret < 0) {
+		dev_err(sdw->cdns.dev, "%s: hdac_bus_eml_sdw_power_up failed: %d\n",
+			__func__, ret);
+		goto out;
+	}
+
+	sdw->cdns.link_up = true;
+out:
+	mutex_unlock(sdw->link_res->shim_lock);
+
+	return ret;
+}
+
+static int intel_link_power_down(struct sdw_intel *sdw)
+{
+	int ret;
+
+	mutex_lock(sdw->link_res->shim_lock);
+
+	sdw->cdns.link_up = false;
```

Comment:
is this on purpose that we want to mark the link are powered down even if the power_down call below fails?

Reply:
I must admit I don't know or don't recall.

It's the same pattern as in the existing intel.c so I just followed without thinking,

The only place where this value is used is in the interrupt thread.

````c
	/* Check if the link is up */
	if (!cdns->link_up)
		return IRQ_NONE;
````

I think it makes sense to use this variable as a sign that you cannot deal with an interrupt while you are powering down.

It still sounds rather racy to me.

@bardliao do you remember anything about this?

Reply:
@plbossart @ranj063 IIRC, we set `sdw->cdns.link_up = false;` before actual power down and don't set it back even if fail to power down on purpose. 
See the comment below.
```
                        /*
                         * we leave the sdw->cdns.link_up flag as false since we've disabled
                         * the link at this point and cannot handle interrupts any longer.
                         */
```
That is to avoid superfluous interrupt handling.

Reply:
@ranj063 ok with no change?